test/js/acme.js

```diff
@@ -0,0 +1,89 @@
+var colors = require("colors");
+var crypto = require("./crypto-util");
+var request = require('request');
+function Acme(privateKey) {
+  this.privateKey = privateKey;
+  this.nonces = [];
+}
+
+Acme.prototype.getNonce = function(url, callback) {
+  var req = request.head({
+    url: url,
+  }, function(error, response, body) {
+    if (error) {
+      console.log(error);
+      process.exit(1);
+    }
+    if (response && "replay-nonce" in response.headers) {
+      console.log("Storing nonce: " + response.headers["replay-nonce"]);
+      this.nonces.push(response.headers["replay-nonce"]);
+      callback();
+      return;
+    }
+
+    console.log("Failed to get nonce for request");
+    process.exit(1);
+  }.bind(this));
+}
+
+Acme.prototype.post = function(url, body, callback) {
+  // Pre-flight with HEAD if we don't have a nonce
+  if (this.nonces.length == 0) {
+    this.getNonce(url, function() {
+      this.post(url, body, callback);
+    }.bind(this))
+    return;
+  }
+
+  console.log("Using nonce: " + this.nonces[0]);
+  var payload = JSON.stringify(body, null, 2);
+  var jws = crypto.generateSignature(this.privateKey,
+                                     new Buffer(payload),
+                                     this.nonces.shift());
+  var signed = JSON.stringify(jws, null, 2);
+
+  console.log('Posting to', url, ':');
+  console.log(signed.green);
+  console.log('Payload:')
+  console.log(payload.blue);
+  var req = request.post({
+    url: url,
+    headers: {
+      'Content-Length': signed.length
+    },
+    encoding: null // Return body as buffer, needed for certificate response
+  }, function(error, response, body) {
+    console.log(("HTTP/1.1 " + response.statusCode).yellow)
+    Object.keys(response.headers).forEach(function(key) {
```

Comment:
Unlikely case but if there's a connection failure/timeout here (`error` is not null) this will throw a TypeError: Cannot read `headers` of undefined & never hit the callback